File: framework/languages/java/samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/handlers/AllocateBingoRoomHandler.java
package systems.zlink.samples.bingo.server.play.handlers;

import java.util.concurrent.CompletionStage;
import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.bingo.shared.contracts.Messages;

@ZLinkHandlerGroup("play")
public final class AllocateBingoRoomHandler
    implements ZLinkRequestHandler<
        Messages.AllocateBingoRoomReq,
        Messages.AllocateBingoRoomRes> {
    private final BingoRoomDirectory rooms;

    public AllocateBingoRoomHandler(BingoRoomDirectory rooms) {
        this.rooms = rooms;
    }

    @Override
    public CompletionStage<Messages.AllocateBingoRoomRes> handleAsync(
        Messages.AllocateBingoRoomReq request,
        ZLinkRequestContext context) {
        return rooms.allocateAsync(request.actorId(), request.mode())
            .thenApply(Messages.AllocateBingoRoomRes::new);
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/handlers/AllocateBingoRoomHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.handlers

import java.util.concurrent.CompletionStage
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.handlers.ZLinkRequest
import systems.zlink.samples.kotlin.bingo.shared.contracts.AllocateBingoRoomReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.AllocateBingoRoomRes

@ZLinkHandlerGroup("play")
class AllocateBingoRoomHandler(
    private val rooms: BingoRoomDirectory,
) {
    @ZLinkRequest(packetName = "AllocateBingoRoomReq")
    fun handleAsync(request: AllocateBingoRoomReq): CompletionStage<AllocateBingoRoomRes> {
        return rooms.allocateAsync(request.actorId, request.mode)
            .thenApply { AllocateBingoRoomRes(it) }
    }
}
